### PR TITLE
Reactive Implementation of AuthorizedClientServiceOAuth2AuthorizedCli…

### PR DESCRIPTION
**Pull request by ankurpathak**
…entManager

ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager is reactive
version of AuthorizedClientServiceOAuth2AuthorizedClientManager

Fixes: gh-7569

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * that is capable of operating outside of a {@code ServerHttpRequest} context,
+ * e.g. in a scheduled/background thread and/or in the service-tier.
+ *
+ * @author Ankur Pathak
+ * @see ReactiveOAuth2AuthorizedClientManager
+ * @see ReactiveOAuth2AuthorizedClientProvider
+ * @see ReactiveOAuth2AuthorizedClientService
+ * @since 5.3
+ */
+public final class ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager implements ReactiveOAuth2AuthorizedClientManager {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Let's rename to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	private final ReactiveClientRegistrationRepository reactiveClientRegistrationRepository;
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Please remove `reactive` prefix for instance variables. See `DefaultReactiveOAuth2AuthorizedClientManager` for naming conventions.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	private final ReactiveOAuth2AuthorizedClientService reactiveAuthorizedClientService;
+	private ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider = context -> null;
+	private Function<OAuth2AuthorizeRequest, Map<String, Object>> contextAttributesMapper = new DefaultContextAttributesMapper();
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: The type should be `Function<OAuth2AuthorizeRequest, Mono<Map<String, Object>>>` (return `Mono`). See `DefaultReactiveOAuth2AuthorizedClientManager.contextAttributesMapper`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+ * @author Ankur Pathak
+ * @see ReactiveOAuth2AuthorizedClientManager
+ * @see ReactiveOAuth2AuthorizedClientProvider
+ * @see ReactiveOAuth2AuthorizedClientService
+ * @since 5.3
+ */
+public final class ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager implements ReactiveOAuth2AuthorizedClientManager {
+	private final ReactiveClientRegistrationRepository reactiveClientRegistrationRepository;
+	private final ReactiveOAuth2AuthorizedClientService reactiveAuthorizedClientService;
+	private ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider = context -> null;
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Update to `context -> Mono.empty()`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+				);
+		// @formatter:on
+	}
+
+	/**
+	 * Sets the {@link ReactiveOAuth2AuthorizedClientProvider} used for authorizing (or re-authorizing) an OAuth 2.0 Client.
+	 *
+	 * @param reactiveAuthorizedClientProvider the {@link ReactiveOAuth2AuthorizedClientProvider} used for authorizing (or re-authorizing) an OAuth 2.0 Client
+	 */
+	public void setReactiveOAuth2AuthorizedClientProvider(ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Rename to `setAuthorizedClientProvider`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+				.map(contextBuilder -> contextBuilder.principal(principal)
+											.attributes(attributes -> {
+												Map<String, Object> contextAttributes = this.contextAttributesMapper.apply(authorizeRequest);
+												if (!CollectionUtils.isEmpty(contextAttributes)) {
+													attributes.putAll(contextAttributes);
+												}
+											})
+											.build()
+				).flatMap(authorizationContext -> this.reactiveAuthorizedClientProvider.authorize(authorizationContext)
+						.doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Rename `x` to `authorizedClient`

**Comment by ghostd** [comment]
Hi,

(not sure where i should post this comment... so i put it here)

I had to change some code of the class to make it work with my use case.

# How i tested this PR (maybe wrongly?)

(in my code, i already renamed `ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`)
```java
    ReactiveClientRegistrationRepository clientRegistrationRepository = new InMemoryReactiveClientRegistrationRepository(clientRegistration);
    ReactiveOAuth2AuthorizedClientService authorizedClientService = new InMemoryReactiveOAuth2AuthorizedClientService(clientRegistrationRepository);

    OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager authorizedClientManager =
        new OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(
                clientRegistrationRepository,
                authorizedClientService
        );
    authorizedClientManager.setReactiveOAuth2AuthorizedClientProvider(new SpecificPartnerReactiveOAuth2AuthorizedClientProvider());

```

The current implementation does not save the authorized client into the repository.

# The code i changed to make it work

I changed
```java
.doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
```
to
```java
.flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))
```

Regards

**Comment by ankurpathak** [comment]
> Hi,
> 
> (not sure where i should post this comment... so i put it here)
> 
> I had to change some code of the class to make it work with my use case.
> 
> # How i tested this PR (maybe wrongly?)
> (in my code, i already renamed `ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`)
> 
> ```java
>     ReactiveClientRegistrationRepository clientRegistrationRepository = new InMemoryReactiveClientRegistrationRepository(clientRegistration);
>     ReactiveOAuth2AuthorizedClientService authorizedClientService = new InMemoryReactiveOAuth2AuthorizedClientService(clientRegistrationRepository);
> 
>     OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager authorizedClientManager =
>         new OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(
>                 clientRegistrationRepository,
>                 authorizedClientService
>         );
>     authorizedClientManager.setReactiveOAuth2AuthorizedClientProvider(new SpecificPartnerReactiveOAuth2AuthorizedClientProvider());
> ```
> 
> The current implementation does not save the authorized client into the repository.
> 
> # The code i changed to make it work
> I changed
> 
> ```java
> .doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
> ```
> 
> to
> 
> ```java
> .flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))
> ```
> 
> Regards

@ghostd I will wait fot @jgrandja to review above commets ad if required will do the needful.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 1 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Resolved

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 2 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Resolved

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 3 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Resolved

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 4 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Resolved

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 5 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Resolved

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 6 above)
Comment (ankurpathak) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Renamed to _authorizedClient as conflicting with local variable authorizedClient.

**Comment by ghostd** [comment]
Hi,

I did some tests with the new version. It seems if `DefaultContextAttributesMapper#apply` returns (a mono of) an empty map, the authentication does not occur. If i change the DefaultContextAttributesMapper to return a non-empty map, the authentication succeeds.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java (hunk 1 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Apologies. I meant to say `AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` - without `OAuth2` prefix so the naming is consistent with existing `AuthorizedClientServiceOAuth2AuthorizedClientManager`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+							)
+				)
+				.flatMap(contextBuilder -> this.contextAttributesMapper.apply(authorizeRequest)
+						.filter(contextAttributes-> !CollectionUtils.isEmpty(contextAttributes))
+						.map(contextAttributes -> contextBuilder.principal(principal)
+								.attributes(attributes -> {
+									attributes.putAll(contextAttributes);
+								}).build())
+				).flatMap(authorizationContext -> this.authorizedClientProvider.authorize(authorizationContext)
+						.doOnNext(_authorizedClient -> authorizedClientService.saveAuthorizedClient(_authorizedClient, principal))
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: @ghostd Thanks for catching this...

> `.flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))`

@ankurpathak This bug was fixed in this [commit](https://github.com/spring-projects/spring-security/commit/8584b12c8da9b73b2579df91b81ff86a49e4daa0#diff-b52f0a944955db35cc1072fdbb4393d2R108).

Can you also apply the update to the test to make use of `PublisherProbe<Void> saveAuthorizedClientProbe` to ensure we cover the scenario of the Mono being subscribed to. See the commit that adds this additional test

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	/**
+	 * The default implementation of the {@link #setContextAttributesMapper(Function) contextAttributesMapper}.
+	 */
+	public static class DefaultContextAttributesMapper implements Function<OAuth2AuthorizeRequest, Mono<Map<String, Object>>> {
+
+		@Override
+		public Mono<Map<String, Object>> apply(OAuth2AuthorizeRequest authorizeRequest) {
+			ServerWebExchange serverWebExchange = authorizeRequest.getAttribute(ServerWebExchange.class.getName());
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: `ServerWebExchange` won't be available in this scenario so `.defaultIfEmpty(Collections.emptyMap())` will execute every time. The logic should be similar to `AuthorizedClientServiceOAuth2AuthorizedClientManager.DefaultContextAttributesMapper`. Looks like a test is missing for this scenario, can you please add.

Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java]: Also see this [comment](https://github.com/spring-projects/spring-security/pull/7589#issuecomment-559243184) as the logic for `.flatMap(contextBuilder -> this.contextAttributesMapper.apply(authorizeRequest)` is incorrect. Please add a test for this as well. If `contextAttributesMapper` returns an empty `Map`, `authorizedClientProvider.authorize()` should still be called.

**Comment by philsttr** [comment]
Any chance this could be added in 5.2.x ?

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManagerTests.java [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+		when(this.authorizedClientService.loadAuthorizedClient(
+				any(), any())).thenReturn(Mono.empty());
+
+		when(authorizedClientProvider.authorize(any())).thenReturn(Mono.empty());
+		OAuth2AuthorizeRequest authorizeRequest = OAuth2AuthorizeRequest.withClientRegistrationId(this.clientRegistration.getRegistrationId())
+				.principal(this.principal)
+				.build();
+		Mono<OAuth2AuthorizedClient> authorizedClient = this.authorizedClientManager.authorize(authorizeRequest);
+
+		authorizedClient.subscribe();
```
Comment (philsttr) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManagerTests.java]: Don't `subscribe()` here.  Instead, rely on `StepVerifier` subscribing (during `verifyComplete()`).   Move the verifications and assertions after the StepVerifier.

The same comment applies to all of your tests where you explicitly call .subscribe()

**Comment by ankurpathak** [comment]
@jgrandja you can takeover this as I tried a lo on this but can't make it.

**Comment by philsttr** [comment]
I can pick this up to continue it.

**Comment by jgrandja** [comment]
@ankurpathak No worries and thank you for your work thus far.

@philsttr That would be great if you can take over this. Much appreciated.

**Comment by philsttr** [comment]
Followup PR created as #7702 

Please close this PR (#7589) in favor of #7702

**Comment by jgrandja** [comment]
Closing in favour of #7702

**Pull request closed without merging**